source/CAS/Client.php

```diff
@@ -1074,6 +1086,145 @@ public function getChangeSessionID()
         return $this->_change_session_id;
     }
 
+    /**
+     * Set the session handler.
+     *
+     * @param SessionHandlerInterface $sessionHandler
+     *
+     * @return bool
+     */
+    public function setSessionHandler(SessionHandlerInterface $sessionHandler)
+    {
+        $this->_sessionHandler = $sessionHandler;
+        return session_set_save_handler($this->_sessionHandler, true);
+    }
+
+    /**
+     * Get a session value using the given key.
+     *
+     * @param string $key
+     *
+     * @return mixed
+     */
+    public function getSessionValue($key)
+    {
+        $this->validateSession($key);
+
+        if (isset($_SESSION[static::PHPCAS_SESSION_PREFIX][$key])) {
+            return $_SESSION[static::PHPCAS_SESSION_PREFIX][$key];
+        }
+
+        return null;
+    }
+
+    /**
+     * Determine whether a session value is set or not.
+     *
+     * @param string $key
+     *
+     * @return bool
+     */
+    public function hasSessionValue($key)
+    {
+        $this->validateSession($key);
+
+        return (isset($_SESSION[static::PHPCAS_SESSION_PREFIX][$key])
+        && !empty($_SESSION[static::PHPCAS_SESSION_PREFIX][$key]));
+    }
+
+    /**
+     * Set a session value using the given key and value.
+     *
+     * @param string $key
+     * @param mixed $value
+     *
+     * @return string
+     */
+    public function setSessionValue($key, $value)
+    {
+        $this->validateSession($key);
+
+        $_SESSION[static::PHPCAS_SESSION_PREFIX][$key] = $value;
+    }
+
+    /**
+     * Remove a session value with the given key.
+     *
+     * @param string $key
+     */
+    public function removeSessionValue($key)
+    {
+        $this->validateSession($key);
+
+        if (isset($_SESSION[static::PHPCAS_SESSION_PREFIX][$key])) {
+            unset($_SESSION[static::PHPCAS_SESSION_PREFIX][$key]);
+        }
+    }
+
+    /**
+     * Remove all phpCAS session values.
+     */
+    public function clearSessionValues()
+    {
+        unset($_SESSION[static::PHPCAS_SESSION_PREFIX]);
+    }
+
+    /**
+     * Ensure phpCAS session value is an array and ensure $key is a string.
+     *
+     * @param string $key
+     *
+     * @return bool
+     */
+    protected function validateSession($key)
+    {
+        if (!is_string($key)) {
+            throw new InvalidArgumentException('Session key must be a string.');
+        }
+
+        if (!isset($_SESSION[static::PHPCAS_SESSION_PREFIX])
+        || !is_array($_SESSION[static::PHPCAS_SESSION_PREFIX])) {
+            $_SESSION[static::PHPCAS_SESSION_PREFIX] = array();
+        }
```

Comment:
Is this validation needed for every session-related operation? I feel this might be too much...

Just figured out that almost every time CAS_Client is initiated, session will be started. We can probably move this to `CAS_Client.__construct()`?

Reply:
What you're suggesting makes sense to me. Is the validation required on every session-related operation? Not sure. If I remember correctly, I think I was just aiming for consistency with existing operation in this PR to keep it easier to review and avoid any big changes in the logic.

We were implementing this into a Laravel project, but basically need to maintain two separate sessions because of how this package is working (Laravel was implementing it's own session management via symfony/session at that time). And implementing this change would've allowed us to integrate it with separate session management methods other than whatever is built into PHP.